src/datalogging.cpp

```diff
@@ -1,108 +1,170 @@
-Adafruit_SPIFlash flash(&flashTransport);
-FatFileSystem fatfs;
+SdFs sd;
 
-void loadSettings() {}
+static uint32_t readUintFromFile(const char* path, uint32_t fallback = 0) {
+    FsFile f = sd.open(path, O_READ);
+    if (!f) return fallback;
+    char buf[32] = {0};
+    int n = f.read(buf, sizeof(buf) - 1);
+    f.close();
+    if (n <= 0) return fallback;
+    return (uint32_t)strtoul(buf, nullptr, 10);
+}
 
-void saveSettings() {}
+static bool writeUintToFile(const char* path, uint32_t v) {
+    FsFile f = sd.open(path, O_WRITE | O_CREAT | O_TRUNC);
+    if (!f) return false;
+    f.println(v);
+    f.close();
+    return true;
+}
+
+static bool appendCsvHeaderIfNew(const char* path) {
+    if (!sd.exists(path)) {
+        FsFile f = sd.open(path, O_WRITE | O_CREAT | O_TRUNC);
+        if (!f) return false;
+        f.println("UptimeMillis,Time,Latitude,Longitude,GPSAltitude,BaroAltitude,Pressure,Temperature,Humidity");
+        f.close();
+    }
+    return true;
+}
 
 void dataloggingInit() {
-    gnssQueue = xQueueCreate(4, sizeof(GnssData));
-    sensingQueue = xQueueCreate(4, sizeof(SensingData));
+    Serial.println("=============");
+    Serial.println(" Datalogging ");
+    Serial.println("=============");
+
+    pinMode(MISO, INPUT_PULLUP);
+    pinMode(SD_CS_PIN, OUTPUT);
+    digitalWrite(SD_CS_PIN, HIGH);
+    SPI.begin();
+    SPI.beginTransaction(SPISettings(1000000, MSBFIRST, SPI_MODE0));
+    for (int i = 0; i < 10; ++i) SPI.transfer(0xFF);
+    SPI.endTransaction();
+    Serial.println("SPI initialized");
+
+    SdSpiConfig cfg(SD_CS_PIN, DEDICATED_SPI, SD_SCK_MHZ(INIT_MHZ));
 
-    if (!fatfs.begin(&flash)) {
-        Serial.println("Failed to initialize filesystem!");
+    if (!sd.cardBegin(cfg)) {
+        Serial.print("\tcardBegin failed  sdErr=0x");
+        Serial.print(sd.sdErrorCode(), HEX);
+        Serial.print(" sdData=0x");
+        Serial.println(sd.sdErrorData(), HEX);
         return;
     }
+    Serial.println("\tCard initialized");
 
-    File file = fatfs.open(BOOTCOUNT_FILE, FILE_READ);
-    if (!file) {
-        bootcount = 1;
-        file = fatfs.open(BOOTCOUNT_FILE, FILE_WRITE);
-        if (file) {
-            file.print(bootcount);
-            file.close();
-        }
+    uint64_t cardSize = sd.card()->cardSize();
+    uint64_t cardSizeBytes = cardSize * 512;
+    Serial.print("\tCard size: ");
+    if (cardSizeBytes < (2ULL * 1024 * 1024)) {
+        Serial.print(cardSizeBytes / 1024);
+        Serial.println(" KB");
+    } else if (cardSizeBytes < (2ULL * 1024 * 1024 * 1024)) {
+        Serial.print(cardSizeBytes / (1024 * 1024));
+        Serial.println(" MB");
     } else {
-        bootcount = file.parseInt() + 1;
-        file.close();
-        file = fatfs.open(BOOTCOUNT_FILE, FILE_WRITE);
-        if (file) {
-            file.print(bootcount);
-            file.close();
-        }
+        Serial.print(cardSizeBytes / (1024 * 1024 * 1024));
+        Serial.println(" GB");
```

Comment:
can u make a function for displaying card size, 10 lines for number formatting in the init is kinda annoying to read through